agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,460 @@
+// For more information, see https://www.consul.io/docs/guides/autopilot.html
+type Autopilot struct {
+	logger   *log.Logger
+	delegate Delegate
+
+	interval       time.Duration
+	healthInterval time.Duration
+
+	clusterHealth     OperatorHealthReply
+	clusterHealthLock sync.RWMutex
+
+	removeDeadCh chan struct{}
+	shutdownCh   chan struct{}
+	waitGroup    sync.WaitGroup
+}
+
+type ServerInfo struct {
+	Name   string
+	ID     string
+	Addr   net.Addr
+	Build  version.Version
+	Status serf.MemberStatus
+}
+
+func NewAutopilot(logger *log.Logger, delegate Delegate, interval, healthInterval time.Duration) *Autopilot {
+	return &Autopilot{
+		logger:         logger,
+		delegate:       delegate,
+		interval:       interval,
+		healthInterval: healthInterval,
+		removeDeadCh:   make(chan struct{}),
+	}
+}
+
+func (a *Autopilot) Start() {
+	a.shutdownCh = make(chan struct{})
+	a.waitGroup = sync.WaitGroup{}
+	a.waitGroup.Add(1)
+
+	go a.run()
+}
+
+func (a *Autopilot) Stop() {
+	close(a.shutdownCh)
+	a.waitGroup.Wait()
+}
+
+// run periodically looks for nonvoting servers to promote and dead servers to remove.
+func (a *Autopilot) run() {
+	defer a.waitGroup.Done()
+
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(a.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case <-ticker.C:
+			autopilotConfig := a.delegate.AutopilotConfig()
+			if autopilotConfig == nil {
+				continue
+			}
+
+			// Skip the non-voter promotions unless all servers support the new APIs
+			minRaftProtocol, err := a.MinRaftProtocol()
+			if err != nil {
+				a.logger.Printf("[ERR] autopilot: error getting server raft protocol versions: %s", err)
+				continue
+			}
+			if minRaftProtocol >= 3 {
+				promotions, err := a.delegate.PromoteNonVoters(autopilotConfig, a.GetClusterHealth())
+				if err != nil {
+					a.logger.Printf("[ERR] autopilot: Error checking for non-voters to promote: %s", err)
+				}
+				if err := a.handlePromotions(promotions); err != nil {
+					a.logger.Printf("[ERR] autopilot: Error handling promotions: %s", err)
+				}
+			}
+
+			if err := a.pruneDeadServers(); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		case <-a.removeDeadCh:
+			autopilotConfig := a.delegate.AutopilotConfig()
+			if autopilotConfig == nil {
+				continue
+			}
+
+			if err := a.pruneDeadServers(); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		}
+	}
+}
+
+// fmtServer prints info about a server in a standard way for logging.
+func fmtServer(server raft.Server) string {
+	return fmt.Sprintf("Server (ID: %q Address: %q)", server.ID, server.Address)
+}
+
+// NumPeers counts the number of voting peers in the given raft config.
+func NumPeers(raftConfig raft.Configuration) int {
+	var numPeers int
+	for _, server := range raftConfig.Servers {
+		if isVoter(server.Suffrage) {
```

Comment:
This is slightly different than the version we had before because it counts `staging` as a voter. Right now the Raft library rips through that state pretty quickly but Diego put that in there to add a built-in autopilot-style promotion in the Raft library itself.

In any event, staging hasn't yet gotten a voter status, so we probably don't want to count it here. Right now `isVoter` is only used for promotion decisions and I think the logic is right there, so we should probably rename it to `isPotentialVoter` to make that more clear.